tfjs-backend-webgpu/src/backend_webgpu.ts

```diff
@@ -623,9 +692,19 @@ export class WebGPUBackend extends KernelBackend {
     const broadcastDimsKey = broadcastDims.map(d => d.join('_')).join(';');
     const key = webgpu_program.makeShaderKey(
         program, bufferShapes, bufferTypes, broadcastDimsKey);
-    const {bindGroupLayout, pipeline} = this.getAndSavePipeline(key, () => {
+    if (program.variableNames.length > 5) {
+      throw new Error(
+          `The length of bindGroupLayouts ` +
+          `${program.variableNames.length + 1}` +
+          `is not supported! Change maxBindGroupLayouts to support it.`);
```

Comment:
Personally, I don't like this change. It makes it non-extensible (for example, `program.variableNames.length` is larger than 5 or we want to use more than one uniform blocks in future). And the benefit is less (I suppose it only saves the first time to create a shader's bind group layout). Is creating bind group layouts a cpu bottleneck for the warmup time? 

@kainino0x take a look. Thanks.

Reply:
If you do want to pre-generate pipeline layouts (I'm not sure whether this is necessary), couldn't it just be done lazily instead of having an error here? i.e. have a method like `getCachedOrCreatePipelineLayout`

I would be somewhat surprised if there was much overhead to creating layout objects, though I don't really know. Internally, IIRC, we cache and deduplicate them. Creation of bind groups is probably more expensive (and could also be potentially reused if that's not done already).

Reply:
I think getCachedOrCreatePipelineLayout is a good idea,  updated, PTAL.

Reply:
>I would be somewhat surprised if there was much overhead to creating layout objects, though I don't really know.

I think the overhead is not much. But the current logic does reduce the number of creating group layout. From this point, the current change looks good to me.

>Creation of bind groups is probably more expensive (and could also be potentially reused if that's not done already).

Yes, you are right. Creation of bind groups is more expensive. But I am not clear about the method of reusing creation of bind groups. Unlike bind group layout, for each bind group, the buffer will change. That's the hard part to cache the bind group.

Reply:
> Unlike bind group layout, for each bind group, the buffer will change. That's the hard part to cache the bind group.

I think there would need to be some heuristics to use the same buffers. For example if running the same model many times (e.g. on a video) it would be nice if we were able to use the same buffers and same bind groups, every frame in the same positions in the model. However making that work seems quite nontrivial.